### PR TITLE
Log information about topic partitions seeing events outside alternative SLA in the EventProducer

### PR DESCRIPTION
**Pull request by somandal**
Log information about topic partitions seeing events outside alternative SLA in the EventProducer
This feature to enable logging such events will be enabled by a config.

### CODE REVIEW

File: datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java [hunk 1]
```diff
@@ -90,11 +96,17 @@
   private final boolean _warnLogLatencyEnabled;
   // Latency threshold at which to log a warning message
   private final long _warnLogLatencyThresholdMs;
+  // Whether to enable logging the list of TopicPartitions with events outside alternative SLA
```
Comment (DEEPTHIKORAT) [inline datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java:99]: nit: alternative --> alternate everywhere.

File: datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java [hunk 2]
```diff
@@ -286,6 +325,10 @@ private void reportMetrics(DatastreamRecordMetadata metadata, long eventsSourceT
       _dynamicMetricsManager.createOrUpdateCounter(MODULE, AGGREGATE, TOTAL_EVENTS_PRODUCED, 1);
       _dynamicMetricsManager.createOrUpdateCounter(MODULE, _datastreamTask.getConnectorType(), TOTAL_EVENTS_PRODUCED,
           1);
+
+      // Log information about events if either warn logging is enabled or logging for topic partitions outside
+      // alternative SLA is enabled
+      performSlaRelatedLogging(metadata, eventsSourceTimestamp, sourceToDestinationLatencyMs);
```
Comment (DEEPTHIKORAT) [inline datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java:331]: is the move inside of _logger.isDebugEnabled() intentional?

File: datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java [hunk 3]
```diff
@@ -413,4 +456,39 @@ private String getDatastreamName() {
 
     return Collections.unmodifiableList(metrics);
   }
+
+  /**
+   * This class encapsulates the topic and partition information for a given event
+   */
+  static class TopicPartitionInfo {
```
Comment (DEEPTHIKORAT) [inline datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java:463]: Why do we need this special class when kafka has TopicPartition class?

File: datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java (hunk 1 above)
Comment (somandal) [reply datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java:99]: good catch 🤣somehow my brain deduped the two words into one

File: datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java (hunk 2 above)
Comment (somandal) [reply datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java:331]: this is not within the if block, the diff just swallowed the rest of the if block.

File: datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java (hunk 3 above)
Comment (somandal) [reply datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java:463]: yeah, was debating whether to add a dependency on kafka or not for this module so I implemented my own. But you're right, it's just silly code duplication. I'll fix this to use Kafka's TopicPartition.

File: datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java (hunk 2 above)
Comment (DEEPTHIKORAT) [reply datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java:330]: Ah sorry, got confused with the indent.

File: datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java [hunk 4]
```diff
@@ -235,6 +253,31 @@ private void reportSLAMetrics(String topicOrDatastreamName, boolean isWithinSLA,
         outsideSLAValue);
   }
 
+  private void performSlaRelatedLogging(DatastreamRecordMetadata metadata, long eventsSourceTimestamp,
+      long sourceToDestinationLatencyMs) {
+    if (_warnLogLatencyEnabled && (sourceToDestinationLatencyMs > _warnLogLatencyThresholdMs)) {
+      _logger.warn("Source to destination latency {} ms is higher than {} ms, Source Timestamp: {}, Metadata: {}",
+          sourceToDestinationLatencyMs, _warnLogLatencyThresholdMs, eventsSourceTimestamp, metadata.toString());
```
[the remark this answers is not in the document]
Comment (somandal) [reply datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java]: removed

**Pull request merged**